uma/protocol/invoice.py

```diff
@@ -0,0 +1,175 @@
+from dataclasses import dataclass
+from typing import Optional
+
+import bech32
+
+from uma.protocol.counterparty_data import (
+    CounterpartyDataOption,
+    CounterpartyDataOptions,
+)
+from uma.protocol.kyc_status import KycStatus
+from uma.tlv_utils import TLVCodable, ByteCodable
+
+
+@dataclass
+class InvoiceCurrency(TLVCodable):
+    # code is the ISO 4217 (if applicable) currency code (eg. "USD"). For cryptocurrencies, this will  be a ticker
+    # symbol, such as BTC for Bitcoin.
+    code: str
+
+    # name is the full display name of the currency (eg. US Dollars).
+    name: str
+
+    # symbol is the symbol used to represent the currency (eg. $).
+    symbol: str
+
+    def __init__(self, code: str = "", name: str = "", symbol: str = "") -> None:
+        self.code = code
+        self.name = name
+        self.symbol = symbol
+
+    @classmethod
+    def tlv_map(cls) -> dict:
+        return {
+            "code": 0,
+            "name": 1,
+            "symbol": 2,
+        }
+
+
+@dataclass
+class InvoiceCounterpartyDataOptions(ByteCodable):
+    options: CounterpartyDataOptions
+
+    def to_bytes(self) -> bytes:
+        pairs = []
+        for key, value in self.options.items():
+            result = key + ":"
+            if value.mandatory:
+                result += "1"
+            else:
+                result += "0"
+            pairs.append(result)
+        pairs.sort()
+        return ",".join(pairs).encode()
+
+    @classmethod
+    def from_bytes(cls, data: bytes) -> "InvoiceCounterpartyDataOptions":
+        options = {}
+        for pair in data.decode().split(","):
+            key, value = pair.split(":")
+            options[key] = CounterpartyDataOption(mandatory=value == "1")
+        return cls(options)
+
+
+@dataclass
+class Invoice(TLVCodable):
+    # Receiving UMA address
+    receiverUma: str
```

Comment:
Fields should probably be snake_case in python. Is there a way we can do that without making the encoding/decoding a huge pain? We do that with the JSONable class in this sdk and the lightspark-sdk for example.

Reply:
Hmm it shouldn't be a problem as the map function maps the field and tlv type. But I'm surprised that pyre, pylint and black didn't warn me about the snake case after all the trouble I had with them :old-man-yells-at-python: